ligo/token/fa2.mligo

```diff
@@ -0,0 +1,144 @@
+// SPDX-FileCopyrightText: 2021 Arthur Breitman
+// SPDX-License-Identifier: LicenseRef-MIT-Arthur-Breitman
+
+#include "../types.mligo"
+
+// -----------------------------------------------------------------
+// Helper
+// -----------------------------------------------------------------
+
+[@inline]
+let check_position_exist (position_id, store : position_id * storage) : position_id =
+  let valid_position_id =
+      match (Big_map.find_opt position_id store.positions) with
+        | Some _ -> position_id
+        | None -> ([%Michelson ({| { FAILWITH } |} : string * unit -> position_id)]
+              ("FA2_TOKEN_UNDEFINED", ()) : position_id)
+  in valid_position_id
+
+[@inline]
+let check_sender (from_ , store : address * storage): address =
+  if (sender = from_) then from_
+  else
+    let key: operator = { owner = from_; operator = sender} in
+    if Big_map.mem key store.operators then
+      from_
+    else
+     ([%Michelson ({| { FAILWITH } |} : string * unit -> address)]
+        ("FA2_NOT_OPERATOR", ()) : address)
+
+// Combine `debit_from` and `credit_to`, we expect position `amt` to be either 1 or 0.
+[@inline]
+let change_position_owner (amt, position_id, from_, to_, store: nat * position_id * address * address * storage): (ledger * position_by_index_map) =
+  if amt = 0n then
+    (store.ledger, store.positions_by_index) // We allow 0 transfer
+  else if amt = 1n then
+    let ledger = store.ledger in
+    let positions_by_index = store.positions_by_index in
+
+    let old_from_ledger_val = match Big_map.find_opt (from_, position_id) ledger with
+          Some ledger_val -> ledger_val
+        | None -> ([%Michelson ({| { FAILWITH } |} : string * (nat * nat) -> ledger_value)]
+              ("FA2_INSUFFICIENT_BALANCE", (amt, 0n)) : ledger_value) in
+
+    // We remove `from` entry from ledger.
+    let updated_from_ledger_val = match Michelson.is_nat (old_from_ledger_val.bal - amt) with
+          Some new_bal ->
+            if (new_bal <> 0n) then (failwith("impossible") : ledger_value option)
+            else (None : ledger_value option)
+        | None -> (failwith("impossible") : ledger_value option) in
```

Comment:
In theory, this duplicates the check on `amt > 1n` (top-level `else`) block.

Probably we could replace `else if amt = 1n then` with just `else` since types don't allow you to subtract too much anyway. A pleasant bonus would be that you will be able to easily include the necessary arguments to this `FA2_INSUFFICIENT_BALANCE` error.

Reply:
That's right, I didn't think of this, updated it now.